nebula-logger/main/logger-engine/classes/Logger.cls

```diff
@@ -685,60 +750,181 @@ global without sharing class Logger {
+        }
+
+        return logEntryEventBuilder;
+    }
+
+    private static String getSuspendSavingLogSystemMessage() {
+        return 'Logging suspended, ignoring call to saveLog()';
+    }
+
+    private static String getSavingLogSystemMessage(SaveMethod saveMethod) {
+        String savingLogMessage = 'Saving ' + LOG_ENTRIES_BUFFER.size() + ' log entries via ';
+        switch on getCurrentQuiddity() {
+            when ANONYMOUS {
+                savingLogMessage += 'Anonymous Apex';
+            }
+            when Aura, VF {
+                savingLogMessage += 'Aura/Lightning/VF component';
+            }
+            when BATCH_ACS, BATCH_APEX, BATCH_CHUNK_PARALLEL, BATCH_CHUNK_SERIAL {
+                savingLogMessage += 'Apex batch job';
+            }
+            when INVOCABLE_ACTION {
+                savingLogMessage += 'Flow invocation';
+            }
+            when QUEUEABLE {
+                savingLogMessage += 'Apex queueable job';
+            }
+            when REST {
+                savingLogMessage += 'Apex REST callout';
+            }
+            when SCHEDULED {
+                savingLogMessage += 'Scheduled job';
+            }
+            when SYNCHRONOUS {
+                savingLogMessage += 'Apex synchronous DML';
+            }
+            when else {
+                savingLogMessage += getCurrentQuiddity().name();
+            }
+        }
+        savingLogMessage += ', save method is ' + saveMethod.name();
+
+        return savingLogMessage;
     }
 
-    // Inner class for saving log entries via a queuable job (using a singleton pattern)
+    // Inner class for saving log entries via a queuable job
     global class QueueableSaver implements Queueable {
         private Id jobId;
 
         private List<LogEntryEvent__e> logEntryEvents = new List<LogEntryEvent__e>();
 
-        global void enqueueLogEntries(List<LogEntryEvent__e> logEntryEvents) {
-            this.logEntryEvents.addAll(logEntryEvents);
-
-            if (jobId == null) {
-                jobId = System.enqueueJob(this);
-            }
+        private QueueableSaver(List<LogEntryEvent__e> logEntryEvents) {
+            this.logEntryEvents = logEntryEvents;
         }
 
         global void execute(System.QueueableContext queueableContext) {
             EventBus.publish(this.logEntryEvents);
         }
     }
+
+    // Inner class for saving log entries via the REST API (using the current user's session ID)
+    private class RestApiSaver {
+        private final String BASE_URL = System.Url.getSalesforceBaseUrl().toExternalForm() + '/services/data/v51.0';
+        private final String COMPOSITE_ENDPOINT = '/composite/sobjects';
+
+        public void insertRecords(List<SObject> records) {
+            HttpRequest request = new HttpRequest();
+            request.setEndpoint(BASE_URL + COMPOSITE_ENDPOINT);
+            request.setHeader('Authorization', 'Bearer ' + UserInfo.getSessionId());
```

Comment:
since `RestApiSaver` relies on the current user being authenticated - should there be a decision made when `saveLog` method is invoked by the site guest user?

Reply:
Added a null check on session ID + recursion to call saveLog() again, using the EventBus